source/common/json/json_internal.cc

```diff
@@ -0,0 +1,668 @@
+
+  reference operator*() {
+    const char& ch = *(ptr.data);
+    if (ch == '\n') {
+      ptr.handler_->line_number_++;
+    }
+    return ch;
+  }
+
+  JsonContainer ptr;
+};
+
+JsonIterator begin(const JsonContainer& c) {
+  return JsonIterator{JsonContainer(c.data, c.handler_)};
+}
+
+JsonIterator end(const JsonContainer& c) {
+  return JsonIterator{JsonContainer(c.data + strlen(c.data), c.handler_)};
+}
+
+void Field::buildJsonDocument(const Field& field, nlohmann::json& value) {
+  switch (field.type_) {
+  case Type::Array: {
+    for (const auto& element : field.value_.array_value_) {
+      switch (element->type_) {
+      case Type::Array:
+      case Type::Object: {
+        nlohmann::json nested_value;
+        buildJsonDocument(*element, nested_value);
+        value.push_back(nested_value);
+        break;
+      }
+      case Type::Boolean:
+        value.push_back(element->value_.boolean_value_);
+        break;
+      case Type::Double:
+        value.push_back(element->value_.double_value_);
+        break;
+      case Type::Integer:
+        value.push_back(element->value_.integer_value_);
+        break;
+      case Type::Null:
+        value.push_back(nlohmann::json::value_t::null);
+        break;
+      case Type::String:
+        value.push_back(element->value_.string_value_);
+      }
+    }
+    break;
+  }
+  case Type::Object: {
+    for (const auto& item : field.value_.object_value_) {
+      auto name = std::string(item.first);
+
+      switch (item.second->type_) {
+      case Type::Array:
+      case Type::Object: {
+        nlohmann::json nested_value;
+        buildJsonDocument(*item.second, nested_value);
+        value.emplace(name, nested_value);
+        break;
+      }
+      case Type::Boolean:
+        value.emplace(name, item.second->value_.boolean_value_);
+        break;
+      case Type::Double:
+        value.emplace(name, item.second->value_.double_value_);
+        break;
+      case Type::Integer:
+        value.emplace(name, item.second->value_.integer_value_);
+        break;
+      case Type::Null:
+        value.emplace(name, nlohmann::json::value_t::null);
+        break;
+      case Type::String:
+        value.emplace(name, item.second->value_.string_value_);
+        break;
+      }
+    }
+    break;
+  }
+  case Type::Null: {
+    break;
+  }
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+}
+
+nlohmann::json Field::asJsonDocument() const {
+  nlohmann::json j;
+  buildJsonDocument(*this, j);
+  return j;
+}
+
+uint64_t Field::hash() const { return HashUtil::xxHash64(asJsonString()); }
+
+bool Field::getBoolean(const std::string& name) const {
+  checkType(Type::Object);
+  auto value_itr = value_.object_value_.find(name);
+  if (value_itr == value_.object_value_.end() || !value_itr->second->isType(Type::Boolean)) {
+    throw Exception(fmt::format("key '{}' missing or not a boolean from lines {}-{}", name,
```

Comment:
what's our general strategy to avoid this occurring during request processing?

Reply:
There is one usage of `loadFromString` in the core dataplane code here, as noted in the description I will swap this with protobuf's json parser in a future PR
> There is one usage in core to parse metadata in the HeaderFormatter:

These get[String/Boolean] methods that can throw are largely unused, but there are <10 usages in filters that I will replace with those with defaults that don't throw in that PR.